File: Javascript/42. Trapping Rain Water.js
/**
 * Link: https://leetcode.com/problems/trapping-rain-water/
 * Author: Dung Nguyen Tien (Chris)
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    var st = [];
    var i = 0;
    var s_area = 0;
    
    while(i < height.length) {
        var peek = st[st.length - 1];
        if (st.length == 0 || height[peek] >= height[i]) {
            st.push(i++);
        }
        else  {
            var middle = st.pop();
            
            if (st.length == 0) continue;
            
            var left = st[st.length - 1];
            var h = Math.min(height[i], height[left]) - height[middle];
            var w = i - left - 1;
            var area = h * w;
            s_area += area;
        }
    }
    
    console.log(i);
    
    return s_area;
};

File: C#/42. Trapping Rain Water.cs
public class Solution {
    public int Trap(int[] height) {
        int res = 0;
        var st = new Stack<int>();
        int i = 0;
        
        while(i < height.Length) {
            if (st.Count == 0 || height[st.Peek()] >= height[i]) {
                st.Push(i++);
            }
            else {
                int right = height[i];
                int mid = height[st.Pop()];
                
                if(st.Count == 0) continue;
                
                int left = height[st.Peek()];
                
                int h = Math.Min(left, right) - mid;
                int w = i - st.Peek() - 1;
                
                res += h * w;
            }
        }
        
        return res;
    }
}